### PR TITLE
add new --cache= logic for speeding up repeated caidx generation

### PR DESCRIPTION
**Pull request by poettering**
Also contains many other fixes including a fix for #112

### CODE REVIEW

File: src/realloc-buffer.c [hunk 1]
```diff
@@ -353,3 +354,41 @@ int realloc_buffer_write_maybe(ReallocBuffer *b, int fd) {
...
+                void *p;
+                int n;
+
+                p = realloc_buffer_extend(b, m);
+                if (!p)
+                        return -ENOMEM;
+
+                va_start(ap, fmt);
+                n = vsnprintf(p, m, fmt, ap);
+                va_end(ap);
```
Comment (keszybz) [inline src/realloc-buffer.c]: Even with exponential increases, this seems rather wasteful...

Edit: OK, there are two users, and they both fit in 64 bytes, so this is not wasteful. But if additional users appear which routinely use more than 64 bytes, it'll probably make sense to extend the initial buffer size.

File: src/log.h [hunk 2]
```diff
@@ -24,15 +24,15 @@ static inline int log_oom(void) {
 #define assert_se(x)                                                    \
         do {                                                            \
                 if (!(x)) {                                             \
-                        log_error("%s:%d (%s): assertion failed:" #x "\n", \
-                                  __FILE__, __LINE__, __PRETTY_FUNCTION__); \
+                        log_error("%s:%d (%s): assertion failed: %s",   \
+                                  __FILE__, __LINE__, __PRETTY_FUNCTION__, #x); \
```
Comment (keszybz) [inline src/log.h]: If that ever happened, we'd get an error from gcc, so this isn't really necessary, but OK.

File: src/cadecoder.c [hunk 3]
```diff
@@ -3903,6 +3903,30 @@ static int ca_decoder_finalize_child(CaDecoder *d, CaDecoderNode *n, CaDecoderNo
...
+                }
+                if (r < 0)
+                        return -errno;
+
+                r = ca_feature_flags_time_granularity_nsec(d->replay_feature_flags, &granularity);
+                if (r < 0)
+                        return r;
+
+                if (timespec_to_nsec(st.st_mtim) / granularity != read_le64(&child->entry->mtime) / granularity)
+                        return -EOPNOTSUPP;
```
[the remark this answers is not in the document]
Comment (poettering) [reply src/cadecoder.c]: well, we care about reproducibility... But yeah, there's already a TODO list item to add a --best-effort mode, where we warn and proceed if we cannot properly restore metadata...

File: src/log.h (hunk 2 above)
Comment (poettering) [reply src/log.h]: yeah, but i ran into it ;-)

**Comment by poettering** [comment]
Thanks a million for the review! Will fix the nits you pointed out in a follow-up PR!

**Pull request merged**